scripts/scale-test/setup_cluster.py

```diff
@@ -0,0 +1,237 @@
+import json
+import logging
+import subprocess
+from argparse import ArgumentParser, Namespace
+from typing import List
+
+from benchmarks import juju
+from benchmarks.models import Cluster, Unit
+from benchmarks.utils import timeit
+
+APPLICATION = "microk8s-node"
+DEFAULT_CHANNEL = "1.24/stable"
+
+logging.basicConfig(format="[%(levelname)s]: %(message)s", level=logging.INFO)
+
+
+@timeit
+def install_microk8s(model, units: List[Unit], channel=DEFAULT_CHANNEL):
+    configure_proxy(units)
+    reboot_and_wait(model)
+    install_snap(channel)
+    update_etc_hosts(units)
+
+
+def update_etc_hosts(units: List[Unit]):
+    """
+    Add entries in /etc/hosts of all units for each node in the cluster.
+    This is needed as nodes's hostnames need to be resolvable in order for
+    the add-node/join process to work.
+    """
+    logging.info("Adding units hostnames on /etc/hosts")
+    for u in units:
+        cmd = f"echo {u.ip}\t{u.instance_id} >> /etc/hosts"
+        juju.run(cmd, app=APPLICATION).check_returncode()
+
+
+@timeit
+def install_snap(channel: str):
+    logging.info("Installing microk8s on all units")
+    for cmd, check_returncode in [
```

Comment:
We could place all commands in a bash script, push the script on each node and trigger it with juju run.

Reply:
Mh, I don't think there is a way to do `juju scp` to all nodes at the same time... so I would need to iterate all nodes and then a final `juju run -a` to execute it on all nodes. This makes `n + 1` juju commands (being `n` the number of nodes)

On the other side, as is is now it would be just 6 `juju run -a` commands.

WDYT about keeping it as is for now, and try to find a better solution if it becomes a problem? Is there any other alternative?

I was trying something like this but didn't manage to make it work:
```python
command = """"$(cat <<EOF
set -eu
snap install microk8s --classic --channel=$MICROK8S_VERSION
sudo usermod -a -G microk8s ubuntu
sudo chown -f -R ubuntu ~/.kube
microk8s status --wait-ready
EOF
)"
"""
juju.run(command, app=APP_NAME)
```
I will give it another try.

Reply:
seems like I found a way to achieve this with a single command 😄 

```python
    logging.info("Installing microk8s on all units")
    all_commands = [
        f"snap install microk8s --classic --channel={channel}",
        "sudo usermod -a -G microk8s ubuntu",
        "sudo chown -f -R ubuntu ~/.kube",
        "sudo newgrp microk8s",
        "microk8s start",
        "microk8s status --wait-ready",
    ]
    command = ";".join(all_commands)
    juju.run(command, app=APP_NAME).check_returncode()
```